datadriven/src/test_data_reader.rs

```diff
@@ -0,0 +1,166 @@
+    scanner: Enumerate<Lines<'a>>,
+    pub logger: slog::Logger,
+}
+
+impl<'a> TestDataReader<'a> {
+    pub fn new<P>(source_name: P, content: &'a str, logger: &slog::Logger) -> Self
+    where
+        P: AsRef<Path>,
+    {
+        Self {
+            source_name: source_name.as_ref().to_path_buf(),
+            scanner: content.lines().enumerate(),
+            data: TestData::default(),
+            logger: logger.clone(),
+        }
+    }
+
+    pub fn next(&mut self) -> Result<bool> {
+        loop {
+            let line = self.scanner.next();
+            if line.is_none() {
+                return Ok(false);
+            }
+
+            let mut pos = line.unwrap().0;
+            let mut line = line.unwrap().1.to_string();
+
+            // Only (1) comment (2) empty line
+            // are accepted before argument.
+            if line.starts_with('#') || line.is_empty() {
+                // Skip comment lines.
+                continue;
+            }
+
+            // Support wrapping directive lines using \, for example:
+            // build-scalar \
+            // vars(int)
+            while line.ends_with('\\') {
+                line.pop();
+
+                let l = self
+                    .scanner
+                    .next()
+                    .expect("expect argument ends without '\\'")
+                    .1
+                    .trim();
+
+                // skip blank line, if not we might get some whitespace at the end,
+                // without trimming it, parsing line will fails
+                if !l.is_empty() {
+                    // Add a whitespace for separating arguments
+                    line.push(' ');
+                    line.push_str(l);
+                }
+
+                // We need the last line number of argument
+                pos += 1;
+            }
+
+            debug!(self.logger, "argument_after_cleanup: {}", line);
+
+            // Start reading argument
+            // Init data
+            self.data = TestData::default();
+
+            // Save `line` information for error/debug message usage
+            self.data.pos = format!("{} : L{}", self.source_name.as_path().display(), pos + 1);
+
+            let (cmd, cmd_args) = parse_line(&line, &self.logger)?;
+
+            if cmd.is_empty() {
+                bail!("cmd must not be empty");
+            }
+
+            debug!(self.logger, "cmd: {}, cmd_args: {:?}", cmd, cmd_args,);
+
+            self.data.cmd = cmd;
+            self.data.cmd_args = cmd_args;
+
+            let mut buf = String::new();
+            let mut separator = false;
+
+            loop {
+                let line = self.scanner.next();
+                if line.is_none() {
+                    break;
+                }
+
+                let line = line.unwrap().1;
+
+                if line == "----" {
+                    separator = true;
+                    break;
+                } else {
+                    buf.push_str(line.trim());
```

Comment:
I don't think it's the same as trimming at L109. Besides, it's expected to add a new line here.

Reply:
go impl is really strange, that only the first line and the last line will trim whitespace,
I will follow it...